metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1250,8 +1248,7 @@ private void replay(ApiMessage message, Optional<OffsetAndEpoch> snapshotId, lon
                 // NoOpRecord is an empty record and doesn't need to be replayed
                 break;
             case ZK_MIGRATION_STATE_RECORD:
-                featureControl.replay((ZkMigrationStateRecord) message);
-                break;
+                throw new UnsupportedOperationException("ZK migration is no longer supported.");
```

Comment:
When adding a new controller, would it be possible that ZK_MIGRATION_STATE_RECORD records are still in the metadata log from a migration done previously?

Reply:
Why do you prefer keeping the call to `featureControl.replay()`? This seems unnecessary. Also I'm not sure whether logging has much value, at least not at the `info` level, maybe `debug` or `trace` in case you absolutely want to see all records passing.

Reply:
> Why do you prefer keeping the call to featureControl.replay()? 

It is a matter of personal preference. :)

Another reason is that it maintains usage in the production code. Otherwise, it’s easy to misunderstand and assume, "Oh, this generated code has no reference in the production code, so we can remove it."

Reply:
For that I prefer a comment, so it's effectively a no-op rather than a method call. This also allows to remove the `FeatureControlManager.replay(ZkMigrationStateRecord)` method.

Reply:
> For that I prefer a comment

That's okay with me. We can leave comments in the ZkMigrationRecord.json file to remind developers about its usage, even though it has no references in the production code.

Reply:
@frankvicky please address @mimaison's comments, thanks!